src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.Analyzer.cs

```diff
@@ -129,6 +130,72 @@ protected override bool ReadOnlyFieldAllowed()
+            {
+                var selectionOperation = semanticModel.GetOperation(this.SelectionResult.GetContainingScope());
+
+                switch (symbol)
+                {
+                    case ILocalSymbol local:
+                        if (local.NullableAnnotation == NullableAnnotation.Annotated)
+                        {
+                            var references = selectionOperation.DescendantsAndSelf()
+                                .Where(o => o switch
+                                {
+                                    ILocalReferenceOperation localReference => localReference.Local == symbol,
+                                    IAssignmentOperation assignment => assignment.Target is ILocalReferenceOperation localReference && localReference.Local == symbol,
+                                    _ => false
+                                });
+
+                            var flowState = GetFlowStateFromLocations(references, local.NullableAnnotation);
+                            if (flowState != NullableFlowState.None)
+                            {
+                                return local.Type.WithNullability(flowState);
+                            }
+                        }
+
+                        return local.GetTypeWithAnnotatedNullability();
+                    case IParameterSymbol parameter:
+                        if (parameter.NullableAnnotation == NullableAnnotation.Annotated)
+                        {
+                            var references = selectionOperation.DescendantsAndSelf()
+                                .Where(o => o switch
+                                {
+                                    IParameterReferenceOperation reference => reference.Parameter == symbol,
+                                    IAssignmentOperation assignment => assignment.Target is IParameterReferenceOperation reference && reference.Parameter == symbol,
+                                    _ => false
+                                });
+
+                            var flowState = GetFlowStateFromLocations(references, parameter.NullableAnnotation);
+                            if (flowState != NullableFlowState.None)
+                            {
+                                return parameter.Type.WithNullability(flowState);
+                            }
+                        }
+
+                        return parameter.GetTypeWithAnnotatedNullability();
+
+                    default:
+                        return base.GetSymbolType(semanticModel, symbol);
+                }
+
+                NullableFlowState GetFlowStateFromLocations(IEnumerable<IOperation> references, NullableAnnotation annotatedNullability)
+                {
+                    foreach (var reference in references)
+                    {
+                        var typeInfo = semanticModel.GetTypeInfo(reference.Syntax);
```

Comment:
In the assignment case, is this the flow state of the left hand or right hand side? Does it matter?

Reply:
I'm guessing it's left hand, but I'm not really sure. @333fred ?

Reply:
Flow state is always RHS, Annotations are always LHS. If I understand your question.

Reply:
So in the case of assignment, we're looking to see if it ever can be assigned to null. That makes sense then, we want the RHS.

Reply:
I guess the question was "if you ask for the flow state of foo" in "foo = bar", is that the flow state _after_ the assignment to bar (so it would have the same flow state of bar) or is that the flow state _before_ the assignment of bar?